### PR TITLE
support asymmetric signing and verification with opaque keys

### PR DESCRIPTION
**Pull request by mikedanese**
This enables signing with crypto oracles.

### CODE REVIEW

**Comment by csstaub** [comment]
This looks great, thanks @mikedanese!

**Comment by csstaub** [comment]
I'll try to get this merged later today and probably update the README and/or add some examples in the doc for this. Do you want me to tag a release with your changes once it's merged?

**Comment by mikedanese** [comment]
Thanks. I'm in no rush to get this in. I'm going to propose an alternative API today that I think might be cleaner.

**Comment by mikedanese** [comment]
I prefer the API introduced in the 1st commit. It's more flexible. The API I proposed yesterday is here for reference:

https://github.com/square/go-jose/compare/v2...mikedanese:oracle-signer-old?expand=1

I moved the crypto.Signer to an example. I don't think it needs to be exported by this library since now it can be implemented elsewhere.

**Comment by csstaub** [comment]
How about exposing the OpaqueSigner implementation for crypto.Signer in a sub-package? Seems useful to have it available rather than having folks re-implement themselves.

**Comment by mikedanese** [comment]
To test this properly, we'll either have to copy some of the roundtrip testing code in the main package or move it to a testing package. I went with the copy to start with.

**Comment by mikedanese** [comment]
Looked at the tests for cryptosigner package again. The round trip testing is pretty entangled with jose package. I recommend we go with copying. The other options are:
* move this package functionality into the jose package
* do a large refactor of jose package testing including touching randReader to break cylic dependencies.

**Comment by csstaub** [comment]
I'm ok with copying, that sounds fine to me. And thank you for your work on this! Let me know when you think this is ready for merging/you want me to review more closely.

File: cryptosigner/cryptosigner.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"encoding/base64"
+	"io"
+	"math/big"
+
+	"golang.org/x/crypto/ed25519"
+	"gopkg.in/square/go-jose.v2"
+)
+
+// OpaqueSignerFromCryptoSigner creates a OpaqueSigner from a "crypto".Signer
+func OpaqueSignerFromCryptoSigner(s crypto.Signer) jose.OpaqueSigner {
```
Comment (cjcullen) [inline cryptosigner/cryptosigner.go]: This function name is a little unwieldy (especially in the `cryptosigner` pkg, where it takes a `crypto.Signer`)

Maybe just `Opaque(s crypto.Signer) jose.OpaqueSigner`?

File: cryptosigner/cryptosigner.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	pk := &jose.JSONWebKey{
+		Key: s.Public(),
+	}
+	b, err := pk.MarshalJSON()
+	if err != nil {
+		panic(err)
```
Comment (cjcullen) [inline cryptosigner/cryptosigner.go]: Is `panic()` the best we can do? Would there be any value in letting a caller know "hey, your signer had an unmarshallable key"?

File: cryptosigner/cryptosigner.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	case jose.RS512, jose.PS512, jose.ES512:
+		hash = crypto.SHA512
+	default:
+		return nil, jose.ErrUnsupportedAlgorithm
+	}
+
+	var hashed []byte
+	if hash != crypto.Hash(0) {
+		hasher := hash.New()
+		// According to documentation, Write() on hash never fails
```
Comment (cjcullen) [inline cryptosigner/cryptosigner.go]: I believe in the intent of the interface, and I trust the implementations your using, but no interface can guarantee to me that an implementation will never produce an error :).

File: cryptosigner/cryptosigner.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		var b []byte
+		b, err = s.signer.Sign(s.rand, hashed, hash)
+		if err != nil {
+			return nil, err
+		}
+
+		sig := struct {
+			R, S *big.Int
+		}{}
+		_, err = asn1.Unmarshal(b, &sig)
```
Comment (cjcullen) [inline cryptosigner/cryptosigner.go]:
```
if _, err := ... ; err != nil {
```

File: opaque.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+// OpaqueSigner is an interface that supports signing payloads with opaque
+// private key(s). Private key operations preformed by implementors may, for
+// example, occur in a hardware module. An OpaqueSigner may rotate signing keys
+// transparently to the user of this interface.
+type OpaqueSigner interface {
+	// Public returns the public key of the current signing key.
+	Public() *JSONWebKey
+	// Algs returns a list of supported signing algorithms.
+	Algs() []SignatureAlgorithm
+	// SingPayload signs a payload with the current signing key using the given
```
Comment (cjcullen) [inline opaque.go]: SignPayload

**Comment by mikedanese** [comment]
@csstaub please take a look

**Comment by cjcullen** [comment]
LGTM

**Comment by csstaub** [comment]
Looks good to me, thanks @mikedanese.

**Pull request merged**